In [1]:
import os, sys
from pathlib import Path
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.preprocessing import MultiLabelBinarizer

# Programmatic repo root resolution
sys.path.insert(0, str(Path.cwd().parent if Path.cwd().name == 'notebooks' else Path.cwd()))
from src.data_prep import load_clean, make_split, find_repo_root
from src.preprocessors import GenreBinarizer

REPO_ROOT = find_repo_root()
DATA_PATH = str(REPO_ROOT / 'data' / 'raw' / 'Movies_Reviews_modified_version1.csv')
os.makedirs(REPO_ROOT / 'results' / 'eda_visualizations', exist_ok=True)

print('Setup complete. Path resolved without Google Drive dependency.')


Setup complete. Path resolved without Google Drive dependency.


In [2]:
# Load clean dataset
df = load_clean(DATA_PATH)
print('Cleaned dataset ingested successfully! Shape:', df.shape)


Cleaned dataset ingested successfully! Shape: (16107, 9)


In [3]:
print('Columns in dataset:', df.columns.tolist())
df[['movie_name', 'genres', 'emotion']].head(3)


Columns in dataset: ['Unnamed: 0', 'Ratings', 'Reviews', 'movie_name', 'Resenhas', 'genres', 'Description', 'emotion', 'cleaned_review']
          movie_name                          genres       emotion
0  Waiting to Exhale  ['Comedy', 'Drama', 'Romance']  anticipation
1  Waiting to Exhale  ['Comedy', 'Drama', 'Romance']  anticipation
2  Waiting to Exhale  ['Comedy', 'Drama', 'Romance']  anticipation


In [4]:
print("Dataset loaded successfully!")
print("Shape:", df.shape)
print(df.columns.tolist())
df.head(3)

Dataset loaded successfully!
Shape: (16107, 9)
['Unnamed: 0', 'Ratings', 'Reviews', 'movie_name', 'Resenhas', 'genres', 'Description', 'emotion', 'cleaned_review']
   Unnamed: 0  Ratings                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              

In [5]:
import warnings
warnings.filterwarnings('ignore')

import pandas as pd
import numpy as np
import ast
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.preprocessing import MultiLabelBinarizer

sns.set(style="whitegrid")
%matplotlib inline

print("Libraries imported successfully!")

Libraries imported successfully!


In [6]:
print("First 10 values of the 'genres' column:")
print(df['genres'].head(10))

print("\nData type of genres column:", df['genres'].dtype)
print("\nTotal rows:", len(df))

First 10 values of the 'genres' column:
0    ['Comedy', 'Drama', 'Romance']
1    ['Comedy', 'Drama', 'Romance']
2    ['Comedy', 'Drama', 'Romance']
3    ['Comedy', 'Drama', 'Romance']
4    ['Comedy', 'Drama', 'Romance']
5    ['Comedy', 'Drama', 'Romance']
6    ['Comedy', 'Drama', 'Romance']
7    ['Comedy', 'Drama', 'Romance']
8                        ['Comedy']
9                        ['Comedy']
Name: genres, dtype: str

Data type of genres column: str

Total rows: 16107


In [7]:
# Parsing the stringified lists using ast.literal_eval
def parse_genres(x):
    try:
        if isinstance(x, str):
            return ast.literal_eval(x)
        else:
            return []
    except:
        return []

df['genres_list'] = df['genres'].apply(parse_genres)

print("Parsing completed!")
print("\nSample after parsing:")
print(df[['genres', 'genres_list']].head(8))

Parsing completed!

Sample after parsing:
                           genres               genres_list
0  ['Comedy', 'Drama', 'Romance']  [Comedy, Drama, Romance]
1  ['Comedy', 'Drama', 'Romance']  [Comedy, Drama, Romance]
2  ['Comedy', 'Drama', 'Romance']  [Comedy, Drama, Romance]
3  ['Comedy', 'Drama', 'Romance']  [Comedy, Drama, Romance]
4  ['Comedy', 'Drama', 'Romance']  [Comedy, Drama, Romance]
5  ['Comedy', 'Drama', 'Romance']  [Comedy, Drama, Romance]
6  ['Comedy', 'Drama', 'Romance']  [Comedy, Drama, Romance]
7  ['Comedy', 'Drama', 'Romance']  [Comedy, Drama, Romance]


In [8]:
# Applying MultiLabelBinarizer
mlb = MultiLabelBinarizer()
genre_encoded = mlb.fit_transform(df['genres_list'])

# Create DataFrame with binary columns
genre_df = pd.DataFrame(genre_encoded, columns=mlb.classes_, index=df.index)

# Add prefix for clarity
genre_df = genre_df.add_prefix('genre_')

print("MultiLabelBinarizer applied successfully!")
print("\nNumber of unique genres found:", len(mlb.classes_))
print("\nEncoded genre columns:")
print(genre_df.columns.tolist())
print("\nShape of encoded matrix:", genre_df.shape)

print("\nFirst 5 rows of encoded genres:")
display(genre_df.head())

MultiLabelBinarizer applied successfully!

Number of unique genres found: 20

Encoded genre columns:
['genre_Action', 'genre_Adventure', 'genre_Animation', 'genre_Comedy', 'genre_Crime', 'genre_Documentary', 'genre_Drama', 'genre_Family', 'genre_Fantasy', 'genre_Foreign', 'genre_History', 'genre_Horror', 'genre_Music', 'genre_Mystery', 'genre_Romance', 'genre_Science Fiction', 'genre_TV Movie', 'genre_Thriller', 'genre_War', 'genre_Western']

Shape of encoded matrix: (16107, 20)

First 5 rows of encoded genres:
   genre_Action  genre_Adventure  ...  genre_War  genre_Western
0             0                0  ...          0              0
1             0                0  ...          0              0
2             0                0  ...          0              0
3             0                0  ...          0              0
4             0                0  ...          0              0

[5 rows x 20 columns]


In [9]:
df_encoded = pd.concat([df, genre_df], axis=1)

print("Final shape after adding encoded genres:", df_encoded.shape)
print("\nNew columns added successfully!")

Final shape after adding encoded genres: (16107, 30)

New columns added successfully!


In [10]:
# Calculate frequency of each genre
genre_counts = genre_df.sum().sort_values(ascending=False)

plt.figure(figsize=(14, 7))
sns.barplot(x=genre_counts.index, y=genre_counts.values, palette="viridis")
plt.title("Genre Distribution in Movie Reviews Dataset", fontsize=16, fontweight='bold')
plt.xlabel("Genre", fontsize=12)
plt.ylabel("Number of Reviews", fontsize=12)
plt.xticks(rotation=45, ha='right')
plt.tight_layout()
plt.show()

print("\nTop 10 most common genres:")
print(genre_counts.head(10))


Top 10 most common genres:
genre_Comedy             7939
genre_Drama              7798
genre_Romance            7625
genre_Thriller           2912
genre_Action             2510
genre_Adventure          1926
genre_Fantasy            1633
genre_Horror             1453
genre_Science Fiction    1383
genre_Family             1345
dtype: int64


In [11]:
# Correlation between genres (how often they appear together)
genre_corr = genre_df.corr()

plt.figure(figsize=(16, 14))
sns.heatmap(genre_corr,
            cmap='coolwarm',
            center=0,
            square=True,
            linewidths=0.4,
            cbar_kws={"shrink": 0.7})
plt.title("Genre Co-occurrence Correlation Heatmap", fontsize=16, fontweight='bold')
plt.tight_layout()
plt.show()

In [12]:
# Save Genre Distribution plot
plt.figure(figsize=(14, 7))
sns.barplot(x=genre_counts.index, y=genre_counts.values, palette="viridis")
plt.title("Genre Distribution in Movie Reviews Dataset", fontsize=16, fontweight='bold')
plt.xlabel("Genre", fontsize=12)
plt.ylabel("Number of Reviews", fontsize=12)
plt.xticks(rotation=45, ha='right')
plt.tight_layout()
plt.savefig(REPO_ROOT / 'results' / 'eda_visualizations' / "member3_genre_distribution.png", dpi=300, bbox_inches='tight')
plt.show()

# Save Co-occurrence Heatmap
plt.figure(figsize=(16, 14))
sns.heatmap(genre_corr, cmap='coolwarm', center=0, square=True, linewidths=0.4, cbar_kws={"shrink": 0.7})
plt.title("Genre Co-occurrence Correlation Heatmap", fontsize=16, fontweight='bold')
plt.tight_layout()
plt.savefig(REPO_ROOT / 'results' / 'eda_visualizations' / "member3_genre_cooccurrence_heatmap.png", dpi=300, bbox_inches='tight')
plt.show()

print("Both plots have been saved successfully!")

Both plots have been saved successfully!


---
## Final Leak-Free Preprocessing (Member 3)
**Objective:** Demonstrates Member 3's `GenreBinarizer` within a leak-free setup.
- Loads data with `load_clean()` (16,107 reviews across 1,309 movies).
- Creates `train_df` and `test_df` via `make_split()` grouped by `movie_name`.
- Fits `GenreBinarizer` **strictly on `train_df`** to learn the genre vocabulary.
- Transforms both `train_df` and `test_df`, handling any unseen genres in test by producing zero-filled indicators.
- Prints genre frequency distributions and verifies zero unseen-genre collision errors.

In [13]:
# 1. Grouped train/test split
train_df, test_df = make_split(df, test_size=0.2, random_state=42)
print(f"Grouped Train Split: {len(train_df):,} samples ({train_df['movie_name'].nunique():,} movies)")
print(f"Grouped Test Split:  {len(test_df):,} samples ({test_df['movie_name'].nunique():,} movies)")

# 2. Fit GenreBinarizer strictly on train_df
binarizer = GenreBinarizer()
binarizer.fit(train_df["genres"])

train_genre_matrix = binarizer.transform(train_df["genres"])
test_genre_matrix = binarizer.transform(test_df["genres"])

print(f"\nLearned Genre Vocabulary Size: {len(binarizer.genres_)} genres")
print(f"Genre Vocabulary: {binarizer.genres_}")
print(f"Train Genre Matrix Shape: {train_genre_matrix.shape}")
print(f"Test Genre Matrix Shape:  {test_genre_matrix.shape}")

# 3. Frequency table of top genres on train split
train_genre_df = pd.DataFrame(train_genre_matrix, columns=binarizer.get_feature_names_out())
top_genres = train_genre_df.sum().sort_values(ascending=False).head(5)
print("\nTop 5 Genres in Training Split:")
for g, cnt in top_genres.items():
    print(f"   • {g}: {cnt:,} movies ({cnt / len(train_df) * 100:.1f}%)")


Grouped Train Split: 12,886 samples (1,048 movies)
Grouped Test Split:  3,221 samples (261 movies)

Learned Genre Vocabulary Size: 20 genres
Genre Vocabulary: ['action', 'adventure', 'animation', 'comedy', 'crime', 'documentary', 'drama', 'family', 'fantasy', 'foreign', 'history', 'horror', 'music', 'mystery', 'romance', 'science_fiction', 'thriller', 'tv_movie', 'war', 'western']
Train Genre Matrix Shape: (12886, 20)
Test Genre Matrix Shape:  (3221, 20)

Top 5 Genres in Training Split:
   • genre_comedy: 6,289 movies (48.8%)
   • genre_drama: 6,226 movies (48.3%)
   • genre_romance: 6,010 movies (46.6%)
   • genre_thriller: 2,428 movies (18.8%)
   • genre_action: 2,166 movies (16.8%)
